Search-by-TFM Frontend changes

advay26 opened the pull request:
Addresses https://github.com/NuGet/Engineering/issues/4579

This work covers all the frontend changes required to enable Search-by-TFM on NuGet.org:

- Moves all Search Filters to the left of the search results (not a collapsible section anymore)
- Separates 'Sort By', places it in the top right
- Unifies the search bar, the search filters panel and the 'Sort by' dropdown into one form (which fixes this search filters bug: https://github.com/NuGet/NuGetGallery/issues/8660)
- Implements checkbox selection logic, where Framework generation checkboxes and TFM checkboxes depend on each other's state
- Constructs query parameters with the selected filters
- A11y FastPass doesn't flag any issues

Previously,
![image](https://user-images.githubusercontent.com/82980589/211931784-7e5b3c9a-e626-49b5-98a3-91f1357a9208.png)
![image](https://user-images.githubusercontent.com/82980589/211931848-f96bb54f-6500-48c9-8377-37b267bd584a.png)

After the changes,
![image](https://user-images.githubusercontent.com/82980589/212203218-90205f24-a211-414c-9a41-a4fc8696f195.png)

![image](https://user-images.githubusercontent.com/82980589/211052110-a4e4dc38-893a-4bbd-9d7d-5a3665683a9f.png)
![image](https://user-images.githubusercontent.com/82980589/211041395-7aafbdd6-6897-4cf9-b1e8-3f1217b6f16d.png)

**UPDATE:**

* Added a tooltip next to the Framework filter that tells users that this filters package search results based on the target frameworks specified in the package contents.
![image](https://user-images.githubusercontent.com/82980589/212203467-4f293349-797c-4b53-84d8-2cac1014c22e.png)
* Put the Framework filters behind a feature flight. While the feature flight is only available to site admins, regular users will still see the bulk of the UI changes, but will not see the Framework filters themselves. i.e. They will see the filters panel on the left side of the page, the 'Sort by' dropdown in the top-right, etc.
![image](https://user-images.githubusercontent.com/82980589/212203742-95525832-d4e7-48f3-86bb-672141b395e6.png)
* TFMs are also now in descending order so users see more recent TFMs first
![image](https://user-images.githubusercontent.com/82980589/212208715-9af283b7-889b-46fd-94b8-53077ada5359.png)

advay26 commented:
If people would like me to do a live/sync review meeting for these changes where I walk everyone through the different parts of the changes and show how users can interact with the frontend, leave a like on this message, and I'll look to schedule something in the next few days.

lyndaidaii commented:
I think it will be better to include the screenshot about search page without your changesi in description. It's easier for reviewers to review and compare changes before and after.

lyndaidaii commented on src/NuGetGallery/ViewModels/ListPackageItemViewModel.cs:
```diff
@@ -11,7 +11,7 @@ namespace NuGetGallery
     public class ListPackageItemViewModel : PackageViewModel
     {
-        private const int _descriptionLengthLimit = 300;
+        private const int _descriptionLengthLimit = 210;
```
just wondering what's intention change limit to 210?

camigthompson replied:
I am wondering the same.

advay26 replied:
The new UI has the Filters panel on the left side of the search results, which means that the search results column is now narrower. In order to keep the package descriptions to about 2 lines (same as the earlier design), I had to cut them short so that they take the same vertical space. If this is not desired, I can increase this back to 300, or something that allows the description to take up 3 lines instead of 2.

Earlier, 
![image](https://user-images.githubusercontent.com/82980589/212204322-ef49b0e2-17ef-455d-a0dc-5afb86710149.png)

Now,
![image](https://user-images.githubusercontent.com/82980589/212204364-40703a62-1496-45e6-b075-1aebf71a51d2.png)

Specifically, the **210** number came from trying different values and seeing what fit best.

camigthompson replied:
Makes sense to me!

agr replied:
Any issues with narrower screens?

advay26 replied:
On narrower screens, the description text does spill onto the next line. It used to do this for the previous search page too, so this change just maintains the same behavior as earlier, but it doesn't dynamically adjust the description length based on the screen width, which is what we would need to do to fix the issue.

Original search page on narrower screens (spills over into a new line),
![image](https://user-images.githubusercontent.com/82980589/212679692-39bba394-6140-472d-958c-7747a247352e.png)

New search page (also spills over),
![image](https://user-images.githubusercontent.com/82980589/212679762-74b2b5ee-81e2-49ae-a12e-57a0171c3a26.png)

agr replied:
How does it look in a vertical smartphone screen?

advay26 replied:
The filter panel is positioned above the search results so it takes up more vertical space in this view. Earlier, the filter panel was collapsible so it didn't take up as much space when it was hidden.

The description lengths are now shorter so they take up less vertical space in this view (6-7 lines earlier, 5 after the changes)

Previously,
![image](https://user-images.githubusercontent.com/82980589/214150726-28d16102-06d4-45b1-be8a-8c38bc721003.png)
![image](https://user-images.githubusercontent.com/82980589/214150022-5f7e12c4-b6a8-4917-a7cd-aa987d55c9c9.png)
![image](https://user-images.githubusercontent.com/82980589/214150106-3ffa47da-c386-4d37-9a65-b706df8dc9a9.png)

With the changes,
![image](https://user-images.githubusercontent.com/82980589/214149838-676ef1f6-1272-4bf5-b4a3-fa654e8c4552.png)
![image](https://user-images.githubusercontent.com/82980589/214149711-2032d70e-f0c5-497c-9652-6d7ee8aba93b.png)

loic-sharma commented:
I'm concerned that the UI doesn't clarify whether the target framework filters are exact (as in the results are packages that have the selected TFMs) or "loose" (as in the results are packages compatible with the selected TFMs). If I'm a newbie working on a .NET 5 app, I might think I need to check .NET 5, .NET 7, and .NET Standard to see all the relevant packages.

Some thoughts:

* Consider using radio buttons instead of checkboxes for the TFM filter. The checkboxes give the user tons of flexibility, which I assume may add confusion.
* Consider adding a tooltip that either details how the feature works, or, links to docs.
* In the search results, below the heading, consider adding a line like `Showing results that are compatible with .NET 5`. This could also link to docs.

lyndaidaii commented on src/NuGetGallery/ViewModels/PackageListViewModel.cs:
```diff
@@ -65,12 +71,103 @@ public class PackageListViewModel
 
         public bool IsAdvancedSearchFlightEnabled { get; set; }
+
+        public Dictionary<string, FrameworkFilterGroup> FrameworkFilters { get; set; }
```
Just thought, it might be better and clear to put those frameworkFilter into a helper class instead of PackageListViewModel class.

advay26 replied:
I've now moved the Framework filters and the TFM lists to a helper class here: https://github.com/NuGet/NuGetGallery/pull/9342/files#diff-50cad345dfd8a1b562bb127499d6fcff351d75e1f781adc7d4d83a673252b65c
(src/NuGetGallery.Core/Frameworks/FrameworkFilterHelper.cs)

I'm also now using Framework constants from our SupportedFrameworks class instead of direct strings for the TFM filters.

lyndaidaii commented on tests/NuGetGallery.Facts/Services/SearchAdaptorFacts.cs:
```diff
@@ -202,10 +218,58 @@ public void IgnoresCaseForSortOrder(SortOrder sortOrder)
+            }
+
+            public static IEnumerable<object[]> FrameworksAndTfmsFilterData => new[]
+{
```
need more space to align

skofman1 commented:
What's our rollout plan for the new UX? It would be good to let a small group of users to try out the UX and get early feedback before we roll it out generally.

anangaur commented:
To @loic-sharma 's comment,
> I'm concerned that the UI doesn't clarify whether the target framework filters are "exact" (as in the results are packages that have the selected TFMs) or "loose" (as in the results are packages compatible with the selected TFMs).

If these are compat (or loose as stated by Loic), would renaming the filter section to `Framework compatibility` help?

/cc: @JonDouglas

JonDouglas commented:
> I'm concerned that the UI doesn't clarify whether the target framework filters are "exact" (as in the results are packages that have the selected TFMs) or "loose" (as in the results are packages compatible with the selected TFMs). If I'm a newbie working on a .NET 5 app, I might think I need to check .NET 5, .NET 7, and .NET Standard to see all the relevant packages.
> 
> Some thoughts:
> 
> * Consider using radio buttons instead of checkboxes for the TFM filter. The checkboxes give the user tons of flexibility, which I assume could cause confusion. Radio buttons guide the user to a single decision.
> * Consider adding a tooltip that either details how the feature works, or, links to docs.
> * In the search results, below the heading, consider adding a line like `Showing results that are compatible with .NET 5`. This could also link to docs.
> 
> This looks amazing though!!!

Great points. We will likely work on computed compatibility later. I think we brainstormed a "toggle" type of control to indicate exact/compat, but will probably be a next iteration type of thing while people learn and give us feedback.

Checkboxes tend to be what's "in" today for filter experiences in many different ecosystems and marketplaces. Valid points on the radioboxes and we will have to learn first before changing.

A tooltip is definitely a nice to have. If we find that users don't find it intuitive enough, we can definitely add one later to explain any "gotchas". Should be fairly straight-forward for now.

A future item we wanted to add is a filter badge when it is applied. That way it would show above the search results. Another added bonus type thing.

TL;DR - 🚢 🐿️ and let's iterate as we continue to get great feedback.

agr commented:
I just realized that the default state of the filter has no indication that it allows all TFMs:
![image](https://user-images.githubusercontent.com/102933/211101478-abe8153f-3d00-4e72-8e63-74003f039249.png)
No checkboxes in Frameworks section would generally imply that we don't want to see neither of .NET, .NET Core, .NET Standard or .NET Framework packages in results, but I presume currently we'll instead see all of them.

agr commented:
Starting with something like this that would expand into the thing above when appropriate checkboxes are set might make things clearer:
![image](https://user-images.githubusercontent.com/102933/211102777-05a0bc1f-71ad-44d7-b7a6-79fbbb7dac0a.png)

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+    function clickFrameworkCheckbox() {
+        this.indeterminate = false;
+
+        if (this.checked == true) {
```
Just `if (this.checked)`?

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        this.indeterminate = false;
+
+        if (this.checked == true) {
+            updateFrameworkFilters("frameworks", this.id, true);
```
Well, you don't even need `if` here.
```
updateFrameworkFilters("frameworks", this.id, this.checked);
```
Should do the same as the whole `if .. else`.

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        framework.checked = false;
+        updateFrameworkFilters("frameworks", framework.id, false);
+
+        if (this.checked == true) {
```
Same as above. Don't compare boolean values.

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        } else {
+            if (checkedCount === 0) {
+                framework.indeterminate = false;
+            }
```
What `framework.indeterminate` ends up with in this branch if `checkedCount` is not `0`?

agr replied:
Should it be
```
framework.indeterminate = checkedCount !== 0;
```
without `if`?

advay26 replied:
I've changed this code to use your suggestion: https://github.com/NuGet/NuGetGallery/pull/9342/files#diff-62befd1077154785ccc2e4ec7952365fafff3c16a7c289e9ea1358f8b2773efaR47

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        }
+        else if (fieldName == "tfms") {
+            searchField = searchForm.tfms;
+        }
```
What if `fieldName` is neither of the above?

agr replied:
How about passing `searchField` into this function as a parameter? Then caller would simply specify `searchForm.frameworks` or `searchForm.tfms` directly as needed.

camigthompson replied:
I agree with @agr

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        }
+        else {
+            searchField.value = searchField.value.replace(frameworkName + ",", "")
+        }
```
Two separate functions with descriptive names (`addFrameworkFilter` and `removeFrameworkFilter` for example) I think would be more readable.

agr replied:
Buut, if you decide to collapse the `if ... else ...` at the place of calling of this function, feel free to leave this as is.

advay26 replied:
I've used your suggestions from earlier so that we don't need to use `if ... else ...` conditions when calling this method, so I ended up leaving the method itself as it is rather than splitting it in 2.

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+            const checkbox = document.getElementById(framework);
+
+            if (checkbox) {
+                checkbox.click();
```
Should it be `checkbox.checked = true`? Wouldn't `click()` invert the state?

advay26 replied:
Here, I actually wanted to make use of the linked clicking logic that was implemented above. So we first 'click' all the framework generations we need to, and then we 'click' all the TFMs we need to.

Let's say some user decided to edit the query string in the URL with values directly, and put 'netframework' in the frameworks parameter, and 'net472' in the tfms parameter. Here, in effect, every package that has 'net472', also has 'netframework' in its framework gens, and since we're just 'ANDing' all the filters, the 'netframework' parameter isn't useful and can be discarded. In this case, when we initialize the checkboxes, the 'netframework' checkbox and the 'net472' checkbox would have been checked together and it would have been in a weird, illogical state that we wouldn't otherwise allow. To make sure this doesn't happen, we initialize the checkboxes by first clicking on all selected framework gens, and then on the TFMs (allowing it to 'uncheck' any framework gens it needs to), so that the query makes sense again and doesn't include any unnecessary filters.

I'm happy to explain further if that doesn't make sense, but the idea here was to avoid that edge case where users decide to edit query params themselves and it gets in a weird state.

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+            if (checkbox) {
+                checkbox.click();
+            }
+        }
```
Please extract this loop as a function and call it twice.
Alternatively, replace with something like
```
inputFrameworkFilterst.map(id => document.getElementById(id)).forEach(checkbox => checkbox.checked = true);
```

camigthompson replied:
Alternatively, iterate through a single list made up of framework and TFM filters

advay26 replied:
For now, I've extracted that into a function and called it twice rather than do it in one list, because I want to run this code on all the Framework checkboxes first, and then on all the 'TFM' checkboxes. I've explained why we would want to do this here, because the more specific TFM filters should override the selected Framework filters, so we want to 'click' on them after we've 'clicked' on all the Frameworks filters.

agr replied:
I still see 2 for loops here.

advay26 replied:
I'd misunderstood which part of the code you were referring to earlier. I've used your suggestion and replaced the for loops with this: 
`inputFrameworkFilters.map(id => document.getElementById(id)).forEach(checkbox => checkbox.click());`
`inputTfmFilters.map(id => document.getElementById(id)).forEach(checkbox => checkbox.click());`

agr commented on src/NuGetGallery/ViewModels/PackageListViewModel.cs:
```diff
@@ -65,12 +71,103 @@ public class PackageListViewModel
+                new FrameworkFilterGroup(
+                    AssetFrameworkHelper.FrameworkGenerationIdentifiers.Net,
+                    AssetFrameworkHelper.FrameworkGenerationDisplayNames.Net,
+                    new List<string> { "net5.0",
```
Please, format lists consistent with the rest of the codebase. It should look like this:
```
new List<string>
{
    "net5.0",
    "net6.0",
    "net7.0",
    "net8.0"
}
```

agr replied:
Also, consider extracting them into static readonly properties of their own and just reference them here.

advay26 replied:
I've fixed the formatting, and put these lists of TFMs in a helper class separately. They're also now using Framework constants rather than strings that I put there myself:
https://github.com/NuGet/NuGetGallery/pull/9342/files#diff-62d4872f1213fbdf7e1a1b65a9e98a04e2b7786b4a5666877c41a7c5eb88b279R81

agr commented on src/NuGetGallery/Views/Shared/ListPackages.cshtml:
```diff
@@ -26,224 +26,271 @@
+                        <div>
+                            <fieldset>
+                                <legend>Options</legend>
+                                <div style="display: flex;">
```
Avoid inline CSS.

agr replied:
Why do we need flexbox here?

advay26 replied:
Not sure, this seems to have been carried over from earlier. I've removed it now.

advay26 replied:
Not sure, this seems to have been carried over from earlier. I've removed it now.

agr commented on src/NuGetGallery/Views/Shared/ListPackages.cshtml:
```diff
@@ -26,224 +26,271 @@
+                            {
+                                if (Model.TotalCount == 1)
+                                {
+                                    <text>1 package </text>
```
Trailing space. There is one in the `else` branch below, too.

agr commented on src/NuGetGallery/Views/Shared/ListPackages.cshtml:
```diff
@@ -26,224 +26,271 @@
+                <div class="row">
+                    <div class="col-md-9">
+                        <h1 tabindex="0">
+                            @if (String.IsNullOrEmpty(Model.SearchTerm))
```
Should it be `string.IsNullOrWhitespace`?

agr commented on src/NuGetGallery/Views/Shared/ListPackages.cshtml:
```diff
@@ -26,224 +26,271 @@
-                    if (e.which === 2) { // Middle-mouse click
-                        emitClickEvent.call(this, e);
-                    }
+    // Used to track how long the user waited before clicking a search selection.
```
Is it only whitespace change below in this file?

advay26 replied:
Yes, it is

camigthompson commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+            updateFrameworkFilters("frameworks", this.id, false);
+        }
+
+        const tfms = document.querySelectorAll('[parent=' + this.id);
+        tfms.forEach((tfm) => {
+            tfm.checked = false;
+            updateFrameworkFilters("tfms", tfm.id, false);
+        });
```
Why do we start with none of the TFMs selected? Don't we want all of them if there are none specified by the user?

advay26 replied:
Since we combine these filters by 'ANDing' them, the idea here is that with every additional checkbox we check, it restricts/shrinks the result set further, so no checkboxes should mean every package is visible. Other framework/platform filtering experiences seem to work similarly too. I'm open to suggestions on how to make this clearer to users, but checking all checkboxes would imply that a package needs to be targeting 'ALL' of those frameworks.

Also, the framework checkboxes we show here don't cover all frameworks available in our ecosystem (we just show the ones we officially support), so even if we were 'ORing' these filters, it wouldn't technically cover all packages.

advay26 replied:
@agr This is in response to your comment here too https://github.com/NuGet/NuGetGallery/pull/9342#issuecomment-1374144302

The default state without checkboxes selected is supposed to indicate that we're not restricting the result set at all, and every additional checkbox we check would make our search more selective. Maybe we can add something more to the tooltip to make this clearer to users?

camigthompson replied:
@advay26 The part about leaving the checkboxes unchecked to start off with makes sense to me, but the part I'm a bit confused about is the ANDing. Why wouldn't we want to show packages that are applicable for at least 1 of the set of checkboxes selected? For example, when I am filtering products on an online store and I select 3 categories (ex: pants, shirts, hats) I want to see all pants AND all shirts AND all hats, not just products that fall into all 3 categories. I would think that if I check net471, net472, and netcoreapp1.0, I want to see packages that are compatible with any combination of 1, 2 or all 3 of those frameworks, or am I misunderstanding this feature?

advay26 replied:
@camigthompson Since package consumers often [target multiple frameworks](https://learn.microsoft.com/en-us/nuget/create-packages/multiple-target-frameworks-project-file#create-a-project-that-supports-multiple-net-framework-versions) in their projects, they would usually want to search for packages that are compatible with **ALL** of these target frameworks.
![image](https://user-images.githubusercontent.com/82980589/212403001-1e71e88a-a22c-4248-961c-3130aa22cb65.png)
From https://learn.microsoft.com/en-us/nuget/consume-packages/finding-and-choosing-packages#determine-supported-frameworks)

We do want to eventually allow greater flexibility to users who want to search for packages compatible with one framework **OR** another, and we will look to add that in later iterations of this (alongside computed framework compatibility vs. exact asset frameworks, etc.). Right now, users can still somewhat approximate this by clicking the framework generation checkbox, which in essence searches for packages that have **ANY** one of the inidividual TFMs within that generation.

camigthompson commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        const framework = document.getElementById(this.getAttribute('parent'));
+        const tfms = document.querySelectorAll('[parent=' + this.getAttribute('parent') + ']');
+
+        let checkedCount = 0;
+        for (const tfm of tfms) {
+            if (tfm.checked) {
+                checkedCount++;
+            }
+        }
+
+        framework.checked = false;
+        updateFrameworkFilters("frameworks", framework.id, false);
+
+        if (this.checked == true) {
+            framework.indeterminate = true;
+            updateFrameworkFilters("tfms", this.id, true);
+        } else {
+            if (checkedCount === 0) {
+                framework.indeterminate = false;
+            }
+            updateFrameworkFilters("tfms", this.id, false);
+        }
+    }
```
What is `this`?

```suggestion
        framework.checked = false; 
        updateFrameworkFilters("frameworks", framework.id, framework.checked);
        
        framework.indeterminate = false;
        for (const tfm of tfms) {
            if (tfm.checked) {
                framework.indeterminate = true;
            }
        }
        framework.indeterminate = this.checked || framework.indeterminate;
        updateFrameworkFilters("tfms", this.id, framework.indeterminate);
    }
```

advay26 replied:
Here, `this` refers to the individual TFM checkbox this event listener is being added to. I've made changes to this method using the suggestions you and Andrei provided: https://github.com/NuGet/NuGetGallery/pull/9342/files#diff-62befd1077154785ccc2e4ec7952365fafff3c16a7c289e9ea1358f8b2773efaR35-R52

camigthompson commented on src/NuGetGallery/ViewModels/PackageListViewModel.cs:
```diff
@@ -65,12 +71,103 @@ public class PackageListViewModel
+                new FrameworkFilterGroup(
+                    AssetFrameworkHelper.FrameworkGenerationIdentifiers.NetCoreApp,
+                    AssetFrameworkHelper.FrameworkGenerationDisplayNames.NetCoreApp,
+                    new List<string> { "netcoreapp1.0",
+                                       "netcoreapp1.1",
+                                       "netcoreapp2.0",
+                                       "netcoreapp2.1",
+                                       "netcoreapp2.2",
+                                       "netcoreapp3.0",
+                                       "netcoreapp3.1", }
+                    ));
```
Formatting. See above.

camigthompson commented on src/NuGetGallery/ViewModels/PackageListViewModel.cs:
```diff
@@ -65,12 +71,103 @@ public class PackageListViewModel
+                new FrameworkFilterGroup(
+                    AssetFrameworkHelper.FrameworkGenerationIdentifiers.NetStandard,
+                    AssetFrameworkHelper.FrameworkGenerationDisplayNames.NetStandard,
+                    new List<string> { "netstandard1.0",
+                                       "netstandard1.1",
+                                       "netstandard1.2",
+                                       "netstandard1.3",
+                                       "netstandard1.4",
+                                       "netstandard1.5",
+                                       "netstandard1.6",
+                                       "netstandard2.0",
+                                       "netstandard2.1" }
+                    ));
```
Formatting. See above.

camigthompson commented on src/NuGetGallery/ViewModels/PackageListViewModel.cs:
```diff
@@ -65,12 +71,103 @@ public class PackageListViewModel
+                new FrameworkFilterGroup(
+                    AssetFrameworkHelper.FrameworkGenerationIdentifiers.NetFramework,
+                    AssetFrameworkHelper.FrameworkGenerationDisplayNames.NetFramework,
+                    new List<string> { "net20",
+                                       "net30",
+                                       "net35",
+                                       "net40",
+                                       "net45",
+                                       "net451",
+                                       "net452",
+                                       "net46",
+                                       "net461",
+                                       "net462",
+                                       "net47",
+                                       "net471",
+                                       "net472",
+                                       "net48",
+                                       "net481" }
+                    ));
```
Formatting. See above.

camigthompson commented on tests/NuGetGallery.Facts/ViewModels/ListPackageItemViewModelFacts.cs:
```diff
@@ -84,7 +84,7 @@ public void LongDescriptionsTruncated()
         public void LongDescriptionsSingleWordTruncatedToLimit()
         {
-            var charLimit = 300;
+            var charLimit = 210;
```
Why the limit decrease?

advay26 replied:
I did this because the filters panel is now on the left of the search results, so the search results column is now narrower than earlier. So that every package search result still takes the same vertical space (2 lines of description text), I had to cut the descriptions shorter, and 210 was a number I arrived at through experimentation. I've explained further here: https://github.com/NuGet/NuGetGallery/pull/9342#discussion_r1068776400

joelverhagen commented on src/NuGetGallery/ViewModels/PackageListViewModel.cs:
```diff
@@ -65,12 +71,103 @@ public class PackageListViewModel
+                                       "netstandard2.1" }
+                    ));
+
+            FrameworkFilters.Add(AssetFrameworkHelper.FrameworkGenerationIdentifiers.NetFramework,
```
Should this hardcoded data model of frameworks be refactored into https://github.com/NuGet/NuGetGallery/blob/4c0074bb3d8a09128994cca3a73661cf502685c4/src/NuGetGallery.Core/Frameworks/SupportedFrameworks.cs?

It seems like we need to reconcile `SupportedFrameworks` with this _somehow_ since they are connected E2E (e.g. the values selectable in the UI must line up with the values populated into the index by the search jobs).

advay26 replied:
I'm now using Framework constants instead of hardcoded strings, and I've placed the lists of TFMs in the SupportedFrameworks class.

joelverhagen commented on src/NuGetGallery/Views/Shared/ListPackages.cshtml:
```diff
@@ -26,224 +26,271 @@
+    <section role="main" class="container main-container page-list-packages">
+        <div class="row clearfix no-margin">
+            <div class="@(Model.IsAdvancedSearchFlightEnabled ? "col-md-3" : "") no-padding" id="filters-column">
+                @if (Model.IsAdvancedSearchFlightEnabled)
```
I think we need a new flight for this. The existing flight is on in PROD so if we deploy like this the feature will be visible until we turn off the flight. If we rollback, we'll need to remember to turn the flight on. etc etc. Easier to have a separate flight name.

advay26 commented:
Thanks for the feedback everyone! I've addressed most of the comments here and added a few other things:

* Added a tooltip next to the Framework filter that tells users that this filters package search results based on the target frameworks specified in the package contents.
![image](https://user-images.githubusercontent.com/82980589/212203467-4f293349-797c-4b53-84d8-2cac1014c22e.png)
* Put the Framework filters behind a feature flight. While the feature flight is only available to site admins, regular users will still see the bulk of the UI changes, but will not see the Framework filters themselves. i.e. They will see the filters panel on the left side of the page, the 'Sort by' dropdown in the top-right, etc.
![image](https://user-images.githubusercontent.com/82980589/212203742-95525832-d4e7-48f3-86bb-672141b395e6.png)
* TFMs are also now in descending order so users see more recent TFMs first
![image](https://user-images.githubusercontent.com/82980589/212208715-9af283b7-889b-46fd-94b8-53077ada5359.png)
* TFM checkboxes don't receive tab focus when their section is collapsed, only when the user clicks the arrow to expand the TFM section

camigthompson commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        }
+
+        framework.checked = false;
+        framework.indeterminate = checkedCount !== 0;
```
Quick question that might be dumb. Why are you using `!==` and not `!=`?

advay26 replied:
Maybe this isn't actually required here, and `!=` would work just fine too, but the `!==` is supposed to be a stricter comparison, and so it seemed safer to use.

agr commented on src/NuGetGallery.Core/Frameworks/FrameworkFilterHelper.cs:
```diff
@@ -0,0 +1,78 @@
+                        AssetFrameworkHelper.FrameworkGenerationIdentifiers.NetCoreApp,
+                        AssetFrameworkHelper.FrameworkGenerationDisplayNames.NetCoreApp,
+                        SupportedFrameworks.TfmFilters.NetCoreAppTfms
+                                                        .Select(f => f.GetShortFolderName())
```
Too much indentation. Just 4 spaces from the line above should be fine.

agr commented on src/NuGetGallery.Core/Frameworks/FrameworkFilterHelper.cs:
```diff
@@ -0,0 +1,78 @@
+            { 
+                AssetFrameworkHelper.FrameworkGenerationIdentifiers.Net,
+                new FrameworkFilterGroup
+                (
```
Please make parentheses location consistent with the code below.
Also, arguments should be indented with 4 spaces.

agr commented on src/NuGetGallery/Scripts/gallery/page-list-packages.js:
```diff
@@ -2,4 +2,132 @@ $(function() {
+        const framework = document.getElementById(this.getAttribute('parent'));
+        const tfms = document.querySelectorAll('[parent=' + this.getAttribute('parent') + ']');
+
+        let checkedCount = 0;
```
How about
```
const checkedCount = tfms.reduce((accumulator, tfm) => accumulator + (tfm.checked ? 1: 0));
```
?

advay26 replied:
I've changed the code to use this.
`const checkedCount = Array.from(tfms).reduce((accumulator, tfm) => accumulator + (tfm.checked ? 1 : 0), 0);`

We need to cast the NodeList to an Array first for this to work. This makes the code more concise, does it also affect efficiency at all?

agr replied:
No clue and unless it is going to be run many times it likely doesn't matter much.

agr reviewed (approved):
LGTM code-wise.
Not a fan of UX, but I guess we'll ask for feedback once it is out.
